Plugins/Public/contraband/Main.cpp

```diff
@@ -0,0 +1,367 @@
+						}
+					}
+				}
+				else if (ini.is_header("Laws"))
+				{
+					while (ini.read_value())
+					{
+						if (ini.is_value("house"))
+						{
+							mapHouseLawList[ini.get_value_string(0)] = stows(ini.get_value_string(1));
+							lawsLoaded++;
+						}
+					}
+				}
+			}
+		}
+		ini.close();
+
+		ConPrint(L"Contraband: Houses Loaded: %u \n", housesLoaded);
+		ConPrint(L"Contraband: House Systems Loaded: %u \n", houseSystemsLoaded);
+		ConPrint(L"Contraband: Laws Loaded: %u \n", lawsLoaded);
+		ConPrint(L"Contraband: Illegal Items Loaded: %u \n", contrabandLoaded);
+	}
+
+	string goods = R"(..\data\equipment\select_equip.ini)";
+	if (ini.open(goods.c_str(), false))
+	{
+		while (ini.read_header())
+		{
+			if (ini.is_header("Commodity"))
+			{
+				uint ids_name;
+				uint nick_id;
+				while (ini.read_value())
+				{
+					if (ini.is_value("nickname"))
+					{
+						nick_id = CreateID(ini.get_value_string());
+					}
+					else if (ini.is_value("ids_name"))
+					{
+						ids_name = ini.get_value_int(0);
+					}
+				}
+				mapItems[nick_id] = ids_name;
+			}
+		}
+		ini.close();
+		ConPrint(L"Contraband: Loaded Items: %u\n", mapItems.size());
+	}
+	HkLoadStringDLLs();
+	
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////
+//Functions
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void ContrabandWarning(uint iClientID, wstring wscText, ...) // WE WANT BOLD RED
+{
+	wchar_t wszBuf[1024 * 8] = L"";
+	va_list marker;
+	va_start(marker, wscText);
+	_vsnwprintf(wszBuf, sizeof(wszBuf) - 1, wscText.c_str(), marker);
+
+	wstring wscXML = L"<TRA data=\"65281\" mask=\"-1\"/><TEXT>" + XMLText(wszBuf) + L"</TEXT>";
+	HkFMsg(iClientID, wscXML);
+}
+
+static void CheckCargo(int iClientID)
+{
+	if (set_lawsContrabandEnabled)
+	{
+		uint iSystemID;
+		pub::Player::GetSystem(iClientID, iSystemID);
+
+		int iHoldSize;
+		list<CARGO_INFO> lstCargo;
+		HkEnumCargo((const wchar_t*)Players.GetActiveCharacterName(iClientID), lstCargo, iHoldSize);
+
+		if (mapHouseAndSystems[iSystemID] != mapHouseAndSystems[iClient_systems[iClientID]] || mapHouseAndSystems.find(iClient_systems[iClientID]) == mapHouseAndSystems.end())
+		{
+			string currentHouse = mapHouseAndSystems[iSystemID];
+			if(mapHouseCargoList.find(currentHouse) != mapHouseCargoList.end())
+			{
+				list<uint> cfgCargoLst = mapHouseCargoList[currentHouse];
+				for (list<CARGO_INFO>::iterator i = lstCargo.begin(); i != lstCargo.end(); ++i)
+				{
+					if(find(cfgCargoLst.begin(), cfgCargoLst.end(), i->iArchID) != cfgCargoLst.end())
+					{
+						wstring illegalCargo = HkGetWStringFromIDS(mapItems[i->iArchID]);
+						ContrabandWarning(iClientID, L"WARNING: You are carrying %s which is illegal in this sector.", illegalCargo.c_str());
+					}
+				}
+			}
+		}
+		iClient_systems[iClientID] = iSystemID; // Update it after everything else is done.
```

Comment:
It's done as a form of validation to ensure that the current system is always up to date. This updates things no matter the outcome of the conditional statements above, and since the check is run whenever some undocks/jumps systems, it ensures that the current system is always up to date.

Reply:
It sounds like we should move that particular line directly into the undock/jump hook handlers, after the call to this function. I know it'll logically be the same but it would make more sense when reading.

Reply:
Understood. Will make the change.

Reply:
How much code duplication?